plugins/gitlab-catalog-backend/src/GitlabUserProcessor.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright SeatGeek
+ * Licensed under the terms of the Apache-2.0 license. See LICENSE file in project root for terms.
+ */
+import { Entity, isUserEntity } from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import { LocationSpec } from '@backstage/plugin-catalog-common';
+import type {
+  CatalogProcessor,
+  CatalogProcessorEmit,
+} from '@backstage/plugin-catalog-node';
+import { ExpandedUserSchema, Gitlab } from '@gitbeaker/rest';
+import { Logger } from 'winston';
+
+const GITLAB_PER_PAGE_LIMIT = 500;
+const GITLAB_DEFAULT_HOST = 'https://gitlab.com';
+
+/**
+ * The GitlabUserProcessor is used to enrich our User entities with information
+ * from Gitlab: notably the user's Gitlab ID.
+ *
+ * @public
+ */
+export class GitlabUserProcessor implements CatalogProcessor {
+  // @ts-ignore: intended to reference as such by the library
+  private readonly gitlab: Gitlab;
+  private readonly logger: Logger;
+  private cacheLoaded: boolean;
+  private userLookup: Map<string, ExpandedUserSchema>;
+  // guarantee that users are loaded only once
+  private loadUserPromise: Promise<Map<string, ExpandedUserSchema>> | null =
+    null;
+
+  private async fetchUsers(): Promise<Map<string, ExpandedUserSchema>> {
+    if (!this.gitlab) {
+      return new Map();
+    }
+
+    if (!this.cacheLoaded) {
+      // we use a shared promise to make sure that the load is only
+      // executed once even if `GitlabUserProcessor.postProcessEntity`
+      // is called "concurrently".
+      if (!this.loadUserPromise) {
+        this.loadUserPromise = this.loadUsers();
+      }
+
+      // Wait for the data to load
+      await this.loadUserPromise;
+    }
+
+    return this.userLookup;
+  }
+
+  private async loadUsers(): Promise<Map<string, ExpandedUserSchema>> {
+    this.logger.info('Loading gitlab users');
+
+    let members: ExpandedUserSchema[] = [];
+    try {
+      members = await this.gitlab!.Users.all({
+        perPage: GITLAB_PER_PAGE_LIMIT,
```

Comment:
does this client paginate on its own?

Reply:
yea, it handles multiple variations of Gitlabs native pagination so we can either allow the client to collect all results before returning, or return one page at a time